Implement name inference for t(), transpose(...)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24474 Implement name inference for torch.dot
* #24471 Implement name inference for addmv, addmv_, mv
* #24469 Implement name inference for expand
* #24306 Implement name inference for mm, addmm, addmm_
* #24921 Fix binary op name inference between unnamed and named tensors.
* #24914 Add helper function Tensor::names()
* #24907 Rename Tensor::names() to Tensor::opt_names()
* #24885 Fix `FIXME_default_names` by storing static list of 64 none names
* #24367 Add thread-local-state NamesMode and NoNamesGuard
* **#24203 Implement name inference for t(), transpose(...)**

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

Differential Revision: [D16915800](https://our.internmc.facebook.com/intern/diff/D16915800)

zou3519 committed:
Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2197,6 +2205,12 @@
   device_guard: False
+  named_guard: False
+
+- func: transpose_(Tensor(a!) self, Dimname dim0, Dimname dim1) -> Tensor(a!)
```
Sorry, let me add tests for in-place transposes before this PR gets reviewed.

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

nairbv commented on test/test_namedtensor.py:
```diff
@@ -272,6 +272,21 @@ def test_stride(self):
+        t = torch.randn(2, 3, names=('N', 'C'))
+        self.assertEqual(t.t().names, ['C', 'N'])
+
+    @unittest.expectedFailure
```
what's expected to fail?

zou3519 replied:
In-place transpose ops. I'll make the test name clearer.

nairbv replied:
fail with an error or the assertions should fail? both fail? I believe a failed assertion raises an exception, so if you're expecting the first assertion to fail I'd think the second one wouldn't get run.

If it's the assertion, I'd think you could assert not equal?

zou3519 replied:
Eh, It might be better to just delete this test. Maintaining it is more trouble than it's worth, and it's meant to be a TODO for me.

nairbv replied:
what is it that doesn't work or fails though? maybe an assertRaises would make sense?

nairbv replied:
I'm fine leaving the test out if it doesn't make sense, I just want to understand what is/isn't supported. wasn't clear to me from the (now removed) test

zou3519 replied:
What isn't supported are t_ and transpose_. So the right way to write the tests would have been to have 2 expected failures, one for each of them.

I don't want to do an assertRaises because that gives the impression that it is intentional that both of them are not supported. It isn't intentional, I just haven't had the time to implement them correctly.

Since the pull request title only talks about t() and transpose(), and not about the in-place variants, the current tests should be sufficient.

nairbv replied:
ah OK.

It's fine to leave the tests out, but I think it is sort of conventional in our codebase to add assertions of failure for features we intend to add later. It sort of points us to and reminds us of where the tests go when we add the functionality later.

zou3519 replied:
>  I think it is sort of conventional in our codebase to add assertions of failure for features we intend to add later.

Sure. However, I'd like to come up with a good plan for doing assertions like that before committing to it. Having all the unimplemented ops spread over multiple tests is not the best, so it would be nice to gather them all in one table or something.

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head

zou3519 committed:
Update on "Implement name inference for t(), transpose(...)"

Implement name inference for t(), transpose(...)

Test Plan
- [namedtensor ci]

gh-metadata: pytorch pytorch 24203 gh/zou3519/103/head
